src/utils/ldap.py

```diff
@@ -242,8 +242,38 @@ def update(self, rdn, attr_dict, new_rdn=False):
 
         """
         dn = rdn + "," + self.basedn
-        actual_entry = self.search(rdn, attrs=None)
-        ldif = modlist.modifyModlist(actual_entry[0], attr_dict, ignore_oldexistent=1)
+        current_entry = self.search(rdn, attrs=None)
+        # Previously, we used modifyModlist, which directly uses the lib system libldap
+        # supplied with openldap. Unfortunately, the output of this command was not
+        # optimal with attributes containing lists (complete deletion then complete
+        # rewriting of the list). In view of the major performance problems associated
+        # with our inherited permissions system, we decided to rewrite this part to
+        # optimize the output.
+        # ldif = modlist.modifyModlist(current_entry[0], attr_dict, ignore_oldexistent=1)
+        ldif = []
+        for attribute, value in attr_dict.items():
+            if attribute in current_entry[0]:
+                current_value = current_entry[0][attribute]
+                if value == current_value:
+                    continue
+
+                # Add or/and delete only needed values in set or list
+                if isinstance(value, (set, list)) and isinstance(current_value, (set, list)):
+                    values_to_add = list(set(value) - set(current_value))
+                    if values_to_add:
+                        ldif.append((ldap.MOD_ADD, attribute, values_to_add))
+                    values_to_del = list(set(current_value) - set(value))
+                    if values_to_del:
+                        ldif.append((ldap.MOD_DELETE, attribute, values_to_del))
+                else:
+                    # Use MOD_REPLACE instead of MOD_DELETE and next MOD_ADD
+                    if isinstance(value, set):
```

Comment:
This test is useless (python will do it anyways when casting), you can just pass `list(value)` to the ldif method. Same line 275.